### ISSUE TITLE
Create 'AssetError' and test 'select_asset' function

### ISSUE THREAD

**Issue by chshersh** [chshersh/tool-sync#83]
This function selects an asset to download based on config:

https://github.com/chshersh/tool-sync/blob/b79eeb91cfdc3a122e6693d503117f68ff1fb44e/src/model/tool.rs#L72-L86

Currently, it returns `Result<Asset, String>` but the goal is to return a custom type for error: `Result<Asset, AssetError`.

The idea is to replace `String` a custom constructor and add tests for these cases.

- [ ] Create a new enum with two constructors
- [ ] Implement the [std::fmt::Display](https://doc.rust-lang.org/std/fmt/trait.Display.html) trait
- [ ] Replace `String` with enum
- [ ] Write tests

**Comment by chshersh** [comment]
Hmm, I didn't know about `Display` 🤔 
Is this the standard and common way to format data types in Rust?

I don't mind using a more standard way of implementing things 🙂 I believe I defined the `display` function manually in several places. So maybe we could start with the refactoring of these functions to use `Display` instead?

```shell
src/config/toml.rs
17:    pub fn display(&self) -> String {

src/model/tool.rs
21:    pub fn display(&self) -> String {

src/sync/archive.rs
28:    pub fn display(&self) -> String {
```

**Comment by MitchellBerend** [comment]
This `Display` trait is like pythons `__repr__` if you know that.

Here is some example:
```rust
use std::fmt;


enum TestEnum {
    Var1(String),
    Var2(String),
}


impl fmt::Display for TestEnum {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            TestEnum::Var1(s) => write!(f, "[VAR1]: {}", s),
            TestEnum::Var2(s) => write!(f, "[VAR2]: {}", s),
        }
    }
}


fn main() {
    let error1 = TestEnum::Var1("error 1".into());
    let error2 = TestEnum::Var2("error 2".into());

    println!("{}", error1);
    eprintln!("{}", error2);
}
```

**Comment by chshersh** [comment]
This looks great! 👏🏻 
In that case, let's go with `std::fmt::Display` instead of our own `display` function. I've edited the issue description.

I like learning about such things about idiomatic Rust code and what are the standard best practices 😌

**Comment by MitchellBerend** [comment]
Im not sure how invasive this change will be but I was thinking that all functions that take some error can be made generic over all types that implement `Display`. That way the error itself can be passed instead of having to convert it manually first.

```rust
pub fn abort_with<Message: Display>(err_msg: Message) -> ! {
    eprintln!(
        r#"Aborting 'tool-sync' with error:

    * {}"#,
        err_msg
    );

    process::exit(1);
}
```
```rust
//These should all work

abort_with("Some non descriptive error");

abort_with(TomlError::IO("The machine is OOM".into()));

```

**Comment by MitchellBerend** [comment]
Id like to leave good-first-issues for others so they can get familiar with the code base. So unless this is blocking some other issue I will leave this open.

**Comment by chshersh** [comment]
@MitchellBerend That's an excellent idea, I completely agree with this 🆗

I've created a separate issue to refactor usages of the existing `display()` method to use the `Display` typeclass:

* https://github.com/chshersh/tool-sync/issues/88

Feel free to create an issue for better `abort_with` functions 👍🏻 
I might create them as well by myself later once #88 is done but you can go ahead and create them straightway.

**Comment by zixuanzhang-x** [comment]
Hello, I'm a Rust beginner and I saw this awesome project! This **development-tools as code** concept is really cool and I'd love to contribute to it!

If there's no one working on this issue already, may I start working on it?

**Comment by MitchellBerend** [comment]
I'm not working on this.

**Comment by chshersh** [comment]
@zixuanzhang-x No one is working so feel free to start 🙂 
Really appreciate contributions 🤗
